docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,239 @@
+state should talk to the application. Otherwise, such tools would have to bake
+in knowledge about Tendermint (e.g., its interfaces and data structures) that
+they might otherwise not need or care about.
+
+However, this raises another issue: The consensus node is the ABCI client, so
+it is inconvenient for the application to "push" work into the consensus
+module. In theory you could work around this by having the consensus module
+"poll" the application for work that needs done, but that has unsatisfactory
+implications for performance and robustness, as well as being harder to
+understand.
+
+There has apparently been some discussion about trying to make a more
+bidirectional communication between the consensus node and the application, but
+this issue seems to still be under discussion. TODO: Impact of ABCI++ for this
+question?
+
+### Context: RPC Issues
+
+The RPC system serves several masters, and has a complex surface area. I
+believe there are some improvements that can be exposed by separating some of
+these concerns.
+
+For light clients, some work is already underway to move toward using P2P
+message passing rather than RPC to synchronize light client state with the rest
+of the network.  Because of the way light client synchronization interacts with
+event logging, this is not as simple as just swapping out the transport: More
+details of the event system will need to be made visible via P2P, and that in
+turn has some implications for fault tolerance (in particular, I think the
+ability of a badly-behaved client to DDoS the network without punishment is a
+concern). However, once this work is done, it should in principle be possible
+to remove the dependency of light clients on the RPC service.
+
+For bootstrapping and operations, the RPC mechanism still makes sense, but can
+be further simplified. Here are some specific proposals:
+
+- Remove the HTTP GET interface entirely.
+- Remove the gRPC interface (this is already planned for v0.36).
+- Remove the websocket interface (TODO: are websockets used anymore?)
+- Simplify the JSON-RPC plumbing to remove unnecessary reflection and interface
+  wrapping.
+
+These changes would preserve the ability of operators to issue queries with
+curl (but would require using JSON-RPC instead of URI parameters). That would
+be a little less user-friendly, but for a use case that should not be that
+prevalent.
+
+These changes would also preserve compatibility with existing JSON-RPC based
+code paths like the `tendermint` CLI and the light client (even ahead of
+further work to remove that dependency).
+
+**Design goal:** An operator should be able to disable non-local access to the
+RPC server on any node in the network without impairing the ability of the
+network to function for service of state replication, including light clients.
+
+**Design principle:** Any communication required to operate the network should use P2P.
```

Comment:
Would this also encompass adding transactions to the validators?

Reply:
Not as I conceive it, but your question is a good one and makes me realize I need to clarify my assumptions: When I say "operate the network" I meant consensus specifically. Operating the application would still need some other mechanism (I'm thinking of this as ABCI, e.g., via gRPC).

Reply:
I added a little more text here to be explicit about that assumption.